### PR TITLE
8314997: Missing optimization opportunities due to missing try_clean_mem_phi() calls

### PR DESCRIPTION
**Pull request by chhagedorn**
While working on a Valhalla bug, I've noticed that we sometimes miss `RegionNode::try_clean_mem_phi()` calls to remove a useless diamond
```
     If
True    False
   Region
```
with only a single memory phi. This blocks further optimizations like converting a loop into a counted one. The code in Valhalla looks slightly different but the problem is also reproducible in mainline.

**Problem**

In the test case, a region is transformed in IGVN such that it merges a diamond without any dependencies on both paths. The region has two phis. One of them is a memory phi which could be transformed by `RegionNode::try_clean_mem_phi()`. But when processing the region with its two phis in IGVN, we do not optimize the memory phi away because `has_unique_phi()` is false and we bail out:
https://github.com/openjdk/jdk/blob/725ec0ce1b463b21cd4c5287cf4ccbee53ec7349/src/hotspot/share/opto/cfgnode.cpp#L450-L471

Later in IGVN, the second phi dies and we only have the single memory phi left. But the region will not be added to the IGVN worklist again to re-apply `try_clean_mem_phi()`. We therefore miss the removal of the diamond and we fail to apply further optimizations. In the test case, we fail to convert the loop into a counted loop.

**Proposed Fix**

The fix I propose is to try to apply `try_clean_mem_phi()` whenever a region is merging a diamond with the assumption that the transformation of a memory phi does not hurt when being applied without being able to remove the region with the diamond (because there are other phis left that cannot be removed). Another option would be to re-add the region to the IGVN worklist when the second last phi dies. But the first approach seems simpler and less invasive.

I've also applied some clean-ups and added an IR test.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314997](https://bugs.openjdk.org/browse/JDK-8314997): Missing optimization opportunities due to missing try_clean_mem_phi() calls (**Enhancement** - P4)

### Reviewers
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [8734fefa](https://git.openjdk.org/jdk/pull/15445/files/8734fefa25047b0bd983ec2eced07d28c011afd0)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15445/head:pull/15445` \
`$ git checkout pull/15445`

Update a local copy of the PR: \
`$ git checkout pull/15445` \
`$ git pull https://git.openjdk.org/jdk.git pull/15445/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15445`

View PR using the GUI difftool: \
`$ git pr show -t 15445`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15445.diff">https://git.openjdk.org/jdk/pull/15445.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15445#issuecomment-1695638334)

### CODE REVIEW

**Review by rwestrel — APPROVED** [review]
Looks reasonable to me.

**Comment by chhagedorn** [comment]
Thanks Roland for your review!

File: src/hotspot/share/opto/cfgnode.cpp [hunk 1]
```diff
@@ -464,32 +464,59 @@ bool RegionNode::try_clean_mem_phi(PhaseGVN *phase) {
...
+bool RegionNode::is_diamond() const {
+  if (req() != 3) {
+    return false;
+  }
+
+  Node* left_path = in(1);
+  Node* right_path = in(2);
+  if (left_path == nullptr || right_path == nullptr) {
+    return false;
+  }
```
Comment (vnkozlov) [inline src/hotspot/share/opto/cfgnode.cpp:503]: So the TOP input will fail next check. May be add comment about that.

File: src/hotspot/share/opto/cfgnode.cpp [hunk 2]
```diff
@@ -1327,42 +1352,60 @@ const Type* PhiNode::Value(PhaseGVN* phase) const {
...
+//          \  /      /    /                       \  /        Some Node
+//         Region    / /-MergeMem     ===>        Region          |
+//          /   \---Phi                             |          MergeMem
+// [other phis]      \                        [other phis]        |
+//                   use                                         use
+bool PhiNode::try_clean_memory_phi(PhaseIterGVN* igvn) {
+  if (_type != Type::MEMORY) {
+    return false;
+  }
+  assert(is_diamond_phi(), "sanity");
```
Comment (vnkozlov) [inline src/hotspot/share/opto/cfgnode.cpp:1389]: Add explicit check `> 0`

**Review by vnkozlov — APPROVED** [review]
Looks good. Only few small comments.

File: src/hotspot/share/opto/cfgnode.cpp (hunk 2 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/cfgnode.cpp]: Good point, added.

File: src/hotspot/share/opto/cfgnode.cpp (hunk 1 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/cfgnode.cpp:503]: I've added a comment at the next `if`.

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review! I've pushed an update.

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by TobiHartmann — APPROVED** [review]
Nice analysis. The fix looks good to me!

**Comment by chhagedorn** [comment]
Thanks Roland, Vladimir, and Tobias for your reviews!

**Pull request closed without merging**